### PR TITLE
Feature/user registration jwt

### PR DESCRIPTION
**Pull request by DandyCV**

### CODE REVIEW

File: app/concepts/api/v1/users/registrations/operation/create.rb [hunk 1]
```diff
@@ -4,6 +4,7 @@ module Api::V1::Users::Registrations::Operation
   class Create < ApplicationOperation
     step :validate_contract
     step :create_user
+    step :send_email
```
Comment (bestwebua) [inline app/concepts/api/v1/users/registrations/operation/create.rb:8]: Rule of thumb: one step - one action. For your case it's: `create_email_token` and `send_email`.

File: app/concepts/application_operation.rb [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 class ApplicationOperation
   include Dry::Transaction
+  include Token
```
Comment (bestwebua) [inline app/concepts/application_operation.rb]: Don't do this. For what in `ApplicationOperation` this behaviour????

File: app/controllers/concerns/token.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Token
```
Comment (bestwebua) [inline app/controllers/concerns/token.rb]: Use singleton instead.

File: app/controllers/concerns/token.rb [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  HMAC_SECRET = Rails.application.credentials.secret_key_base
+  ALGO = 'HS256'
+
+  def encode(payload)
+    JWT.encode payload, HMAC_SECRET, ALGO
```
Comment (bestwebua) [inline app/controllers/concerns/token.rb]: Please use ine code style trought all project.

File: app/controllers/concerns/token.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  end
+
+  def decode(token)
+    JWT.decode token, HMAC_SECRET, true, { algorithm: ALGO }
```
Comment (bestwebua) [inline app/controllers/concerns/token.rb]: The same.

File: app/controllers/concerns/token.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+module Token
+  HMAC_SECRET = Rails.application.credentials.secret_key_base
+  ALGO = 'HS256'
```
Comment (bestwebua) [inline app/controllers/concerns/token.rb]:
```ruby
HMAC = 'HS256'
```

**Review by bestwebua — CHANGES_REQUESTED** [review]

File: app/concepts/application_contract.rb [hunk 7]
```diff
@@ -3,9 +3,9 @@
 class ApplicationContract < Dry::Validation::Contract
   config.messages.default_locale = :en
 
-  def self.call(params, &block)
+  def self.call(params)
     params = params.to_unsafe_hash if params.is_a?((ActionController::Parameters))
```
Comment (bestwebua) [inline app/concepts/application_contract.rb]: Double brackets?? `((ActionController::Parameters))`

File: spec/concepts/api/v1/lib/service/email_token_spec.rb [hunk 8]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+RSpec.describe Api::V1::Lib::Service::EmailToken do
+  let(:payload) { { params: 42 } }
+  let(:hmac_secret) { described_class::HMAC_SECRET }
```
Comment (bestwebua) [inline spec/concepts/api/v1/lib/service/email_token_spec.rb]: For what? Please use stub_const with reference to some let instead, for instance:
```ruby
let(:hmac_secret) { 'hmac_secret' }
```

Or direct call of these ones.

File: spec/concepts/api/v1/lib/service/email_token_spec.rb [hunk 9]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+RSpec.describe Api::V1::Lib::Service::EmailToken do
```
Comment (bestwebua) [inline spec/concepts/api/v1/lib/service/email_token_spec.rb:3]: Where checking for defined constants expectations?

**Review by bestwebua — CHANGES_REQUESTED** [review]

File: app/concepts/application_contract.rb (hunk 7 above)
Comment (DandyCV) [reply app/concepts/application_contract.rb]: Fixed

File: spec/concepts/api/v1/lib/service/email_token_spec.rb (hunk 8 above)
Comment (DandyCV) [reply spec/concepts/api/v1/lib/service/email_token_spec.rb]: Done

File: spec/concepts/api/v1/lib/service/email_token_spec.rb (hunk 9 above)
Comment (DandyCV) [reply spec/concepts/api/v1/lib/service/email_token_spec.rb:3]: Added.

File: spec/concepts/api/v1/lib/service/email_token_spec.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+  let(:payload) { { params: 42 } }
+  let(:hmac_secret) { 'hmac_secret' }
+  let(:hmac) { 'hmac' }
+
+  before do
+    stub_const("#{described_class}::HMAC_SECRET", hmac_secret)
+    stub_const("#{described_class}::HMAC", hmac)
+  end
+
+  it { expect(defined?(described_class::HMAC_SECRET)).to eq('constant') }
```
Comment (bestwebua) [inline spec/concepts/api/v1/lib/service/email_token_spec.rb]: Use native matchers instead.

Comment (DandyCV) [reply spec/concepts/api/v1/lib/service/email_token_spec.rb]: Done.

File: spec/concepts/api/v1/lib/service/email_token_spec.rb [hunk 11]
```diff
@@ -10,8 +10,8 @@
     stub_const("#{described_class}::HMAC", hmac)
   end
 
-  it { expect(defined?(described_class::HMAC_SECRET)).to eq('constant') }
-  it { expect(defined?(described_class::HMAC)).to eq('constant') }
+  it { expect(described_class).to be_const_defined(:HMAC_SECRET) }
```
Comment (bestwebua) [inline spec/concepts/api/v1/lib/service/email_token_spec.rb:13]: 👍

File: app/concepts/api/v1/users/confirmations/operation/index.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+module Api::V1::Users::Confirmation::Operation
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/operation/index.rb]: Should be a `Api::V1::Users::Confirmations::Operation::Show`

File: app/concepts/api/v1/users/confirmations/contract/index.rb [hunk 13]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Api::V1::Users::Registrations::Contract
+  class Create < ApplicationContract
+    params do
+      required(:email_token).filled(:string, min_size?: Constants::Shared::EMAIL_TOKEN_MIN_SIZE)
+      required(:password_confirmation).filled(:string)
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/contract/index.rb]: For what!?

File: app/concepts/api/v1/users/confirmations/contract/index.rb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module Api::V1::Users::Registrations::Contract
+  class Create < ApplicationContract
+    params do
+      required(:email_token).filled(:string, min_size?: Constants::Shared::EMAIL_TOKEN_MIN_SIZE)
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/contract/index.rb]: Needless

File: app/concepts/api/v1/users/confirmations/contract/index.rb [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+      required(:password_confirmation).filled(:string)
+    end
+
+    rule(:email_token) do
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/contract/index.rb]: It's not contract responsibility.

File: app/concepts/api/v1/users/confirmations/operation/index.rb [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    step :validate_data
+    step :update_user
+
+    def validate_token(params)
+      contract = Api::V1::Users::Confirmation::Contract::Index.call(params)
+      contract.success? ? Success(contract) : Failure(contract)
+    end
+
+    def decode_token(contract)
+      Success(Api::V1::Lib::Service::EmailToken.decode(contract.email_token))
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/operation/index.rb]: Email token can be invalid. You can unit 3 first steps.

File: app/concepts/api/v1/users/confirmations/operation/index.rb [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+    end
+
+    def validate_data(data)
+      contract = Api::V1::Users::Confirmation::Contract::Validate.call(data)
+      contract.success? ? Success(contract) : Failure(contract)
+    end
+
+    def update_user(contract)
+      user = User.find(contract.id)
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/operation/index.rb]: What will happen if user doesn't exist?

File: app/concepts/api/v1/users/confirmations/serializer/index.rb [hunk 18]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+module Api::V1::Users::Confirmation::Serializer
+  class Create < ApplicationSerializer
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/serializer/index.rb]: Why you have defined serializer for this endpoint?

**Review by bestwebua — CHANGES_REQUESTED** [review]
Please complete this PR with mailer feature only. Confirmation feature should be in another PR.

File: app/concepts/api/v1/users/confirmations/operation/index.rb (hunk 12 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/confirmations/operation/index.rb]: Renamed

File: app/concepts/api/v1/users/confirmations/contract/index.rb (hunk 13 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/confirmations/contract/index.rb]: This contract is removed

File: app/concepts/api/v1/users/confirmations/contract/index.rb (hunk 14 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/confirmations/contract/index.rb]: This contract is removed

File: app/concepts/api/v1/users/confirmations/contract/index.rb (hunk 15 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/confirmations/contract/index.rb]: This contract is removed

File: app/concepts/api/v1/users/confirmations/operation/index.rb (hunk 16 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/confirmations/operation/index.rb]: This operation will be done in next commits

File: app/concepts/api/v1/users/confirmations/operation/index.rb (hunk 17 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/confirmations/operation/index.rb]: This operation will be done in next commits

File: app/concepts/api/v1/users/confirmations/serializer/index.rb (hunk 18 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/confirmations/serializer/index.rb]: The idea was to show User with new attribute (:confirmed) => true. For now I removed this serializer

File: Gemfile [hunk 19]
```diff
@@ -40,6 +40,8 @@ group :development, :test do
   gem 'pry-byebug', '~> 3.9'
   # Causes rails console to open pry
   gem 'pry-rails', '~> 0.3.9'
+  # Shim to load environment variables from .env into ENV in development
+  gem 'dotenv-rails', '~> 2.7.6'
```
Comment (bestwebua) [inline Gemfile]: Please use out of box rails solutions instead.

File: app/concepts/api/v1/users/confirmations/operation/show.rb [hunk 20]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+module Api::V1::Users::Confirmations::Operation
```
Comment (bestwebua) [inline app/concepts/api/v1/users/confirmations/operation/show.rb]: Please remove from this PR.

File: app/controllers/api/v1/users/confirmations_controller.rb [hunk 21]
```diff
@@ -2,23 +2,7 @@
...
-        end
-
-        result.failure do |failure_object|
-          respond_with(
-            status: 422,
-            entity: failure_object
-          )
-        end
-      end
+    def show
```
Comment (bestwebua) [inline app/controllers/api/v1/users/confirmations_controller.rb]: The same.

File: app/mailers/application_mailer.rb [hunk 22]
```diff
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 
 class ApplicationMailer < ActionMailer::Base
-  default from: 'planner.api.mailer@gmail.com'
+  default from: Constants::Shared::EMAIL_ADDRESS
```
Comment (bestwebua) [inline app/mailers/application_mailer.rb]: `Constants::Shared::DEFAULT_EMAIL` gets you more context. Also more flexible way use value from `Rails.configuration.default_sender_email`, for example.

File: config/routes.rb [hunk 23]
```diff
@@ -4,7 +4,7 @@
   namespace :api do
     namespace :v1 do
       namespace :users do
-        resource :confirmation, only: :index
+        resource :confirmation, only: :show
```
Comment (bestwebua) [inline config/routes.rb]: In another pr.

File: app/mailers/registration_mailer.rb [hunk 24]
```diff
@@ -4,7 +4,7 @@ class RegistrationMailer < ApplicationMailer
   def confirmation_email
```
Comment (bestwebua) [inline app/mailers/registration_mailer.rb]: what about method params?

File: app/mailers/registration_mailer.rb [hunk 25]
```diff
@@ -4,7 +4,7 @@ class RegistrationMailer < ApplicationMailer
...
     user = params[:user]
     @user = user
-    @url = "#{root_url}/api/v1/users/confirmation?email_token=#{params[:email_token]}"
+    @url = "http://#{Rails.application.config.hosts.first}/api/v1/users/confirmation?email_token=#{params[:email_token]}"
```
Comment (bestwebua) [inline app/mailers/registration_mailer.rb]: Please, don't hardcode it. You should think about how to write more flexible.

```ruby
def confirmation(email, token, path)
    @email = email
    @confirmation_url = URI.parse("#{path}?email_token=#{token}").to_s
    mail(to: email, subject: I18n.t('user_mailer.confirmation.subject'))
  end
```

File: config/application.rb [hunk 26]
```diff
@@ -21,6 +21,8 @@
 # you've limited to :test, :development, or :production.
 Bundler.require(*Rails.groups)
 
+Dotenv::Railtie.load
```
Comment (bestwebua) [inline config/application.rb]: Please remove.

File: config/environments/test.rb [hunk 27]
```diff
@@ -48,15 +48,9 @@
 
   # Annotate rendered view with file names.
   # config.action_view.annotate_rendered_view_with_filenames = true
+  config.hosts = [
```
Comment (bestwebua) [inline config/environments/test.rb]: For what?

File: config/environments/test.rb [hunk 28]
```diff
@@ -48,15 +48,9 @@
...
-
-  config.action_mailer.smtp_settings = {
-    :address              => 'smtp.gmail.com',
-    :port                 => 587,
-    :user_name            => Rails.application.credentials.email[:login],
-    :password             => Rails.application.credentials.email[:password],
-    :authentication       => 'plain',
-    :enable_starttls_auto => true
-  }
+  config.action_mailer.default_url_options = { host: config.hosts.first, port: 3000 }
```
Comment (bestwebua) [inline config/environments/test.rb]: config.action_mailer.default_url_options = { host: 'localhost:3000' }
config.default_sender_email = 'no-reply@localhost'

File: spec/mailers/registration_mailer_spec.rb [hunk 29]
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+RSpec.describe RegistrationMailer, type: :mailer do
+  it { expect(described_class).to be < ApplicationMailer }
+
+  describe '#confirmation email' do
```
Comment (bestwebua) [inline spec/mailers/registration_mailer_spec.rb:6]:
```ruby
let(:email) { from_faker... }

  describe '#confirmation' do
    subject(:mailer) { described_class.confirmation(email, token, path) }

    let(:token) { from_faker... }
    let(:path) { from_faker... }

    it 'includes confirmation data' do
      expect(mailer.to).to include(email)
      expect(mailer.from).to include(Rails.configuration.default_sender_email)
      expect(mailer.subject).to eq(I18n.t('user_mailer.confirmation.subject'))
      expect(mailer.body.encoded).to include(URI.parse("#{path}?email_token=#{token}").to_s)
    end
  end
```

**Review by bestwebua — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 19 above)
Comment (DandyCV) [reply Gemfile]: Ok

File: app/concepts/api/v1/users/confirmations/operation/show.rb (hunk 20 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/confirmations/operation/show.rb]: OK

File: app/controllers/api/v1/users/confirmations_controller.rb (hunk 21 above)
Comment (DandyCV) [reply app/controllers/api/v1/users/confirmations_controller.rb]: OK

File: app/mailers/application_mailer.rb (hunk 22 above)
Comment (DandyCV) [reply app/mailers/application_mailer.rb]: Rails.configuration used

File: config/routes.rb (hunk 23 above)
Comment (DandyCV) [reply config/routes.rb]: Ok

File: app/mailers/registration_mailer.rb (hunk 25 above)
Comment (DandyCV) [reply app/mailers/registration_mailer.rb]: Refactored

File: config/application.rb (hunk 26 above)
Comment (DandyCV) [reply config/application.rb]: Removed

File: config/environments/test.rb (hunk 28 above)
Comment (DandyCV) [reply config/environments/test.rb]: config refactored

File: spec/mailers/registration_mailer_spec.rb (hunk 29 above)
Comment (DandyCV) [reply spec/mailers/registration_mailer_spec.rb:6]: 👍

File: app/concepts/api/v1/users/registrations/operation/create.rb [hunk 30]
```diff
@@ -31,7 +31,11 @@ def create_email_token(ctx)
 
     def send_email(ctx)
       user = ctx[:user]
-      RegistrationMailer.with(user: user, email_token: ctx[:email_token]).confirmation_email.deliver_later
+      RegistrationMailer.confirmation_email(
```
Comment (bestwebua) [inline app/concepts/api/v1/users/registrations/operation/create.rb:34]:
```ruby
Success(RegistrationMailer.confirmation_email...
```

File: app/concepts/api/v1/users/confirmations/serializer/index.rb (hunk 18 above)
Comment (bestwebua) [reply app/concepts/api/v1/users/confirmations/serializer/index.rb]: no_content is better solution

File: config/environments/development.rb [hunk 31]
```diff
@@ -56,4 +56,17 @@
 
   # Uncomment if you wish to allow Action Cable access from any origin.
   # config.action_cable.disable_request_forgery_protection = true
+
+  config.action_mailer.default_url_options = { host: 'localhost:3000' }
+
+  config.default_sender_email = 'no-reply@localhost'
+
+  config.action_mailer.smtp_settings = {
```
Comment (bestwebua) [inline config/environments/development.rb]: Use letter_opener for dev env.

File: app/concepts/api/v1/users/registrations/operation/create.rb (hunk 30 above)
Comment (DandyCV) [reply app/concepts/api/v1/users/registrations/operation/create.rb:34]: But controller expects from operation to return User entity, doesn't it?

File: config/environments/development.rb (hunk 31 above)
Comment (DandyCV) [reply config/environments/development.rb]: Added.

**Review by bestwebua — COMMENTED** [review]
👍

**Review by bestwebua — APPROVED** [review]
Done.

**Pull request merged**